src/cljs/athens/components/components_util.cljs

```diff
@@ -0,0 +1,25 @@
+(ns athens.components.components-util
```

Comment:
Not only did I mean to use `cljs` instead `cljc`, I meant that all 3 files can be in one components file. I prefer monolithic files rather than fragmenting across several files, until it's absolutely necessary. See how big `events` and `db` are!

https://github.com/athensresearch/athens/pull/274#discussion_r456837520

Reply:
I'd say that putting components in different files could be more helpful, because the component system should be designed as extensible as possible, especially since we plan to add very complex ones later (such as `query` & `kanban`).

Reply:
Whether the files are distributed or in one file doesn't affect their extensibility. I don't think we will necessarily copy verbatim how Roam implements `{{kanban}}` and `{{query}}` either. I like to build up then break out. Will still be a while until these reach the level of complexity of say `blocks`. Please couple the components for now.

Reply:
I don't mind if the files are together or seperate personally, but putting them all together would make other contributors spend more effort when trying to add their own component (for example, I had to read the entire `blocks` before adding the auto page-link feature), which might discourage collaboration. For now, I can combine them and add separate docs to indicate which part of the file does what, which could hopefully mitigate this issue. `components_util` should remain separate though, because I'm looking into metaprogramming and would soon experiment with ways users could add their own components, and adding them into components declaration would only confuse people.

Reply:
I think if a contributor wanted to add a component, they would look through the previous components to see how it's done. How else would they know how to make a component? If you hadn't read the blocks file, would you have known how to make the auto-page link feature? If someone hasn't read the relevant parts of the codebase, then yes I am would discourage them from contributing yet, if they are somehow even able to complete the feature without prerequisite knowledge.

Maybe a big file is intimidating, but it's tractable. Having to look through more files in a new codebase is worse because 1) you have to know the file exists and 2) you may need to read it anyways, we wouldn't know ahead of time.

I don't know what you intend to do with metaprogramming so I don't know if it would confuse people or not having it multiple files. Please elaborate before you try to use that as support for what you want to do. Like kanban, query, and backends (on Discord), I believe you're assuming too much.

This convo sounds nitpicky. But part of what I'm saying comes from my experience doing DevOps, where I created and ran our company's CI builds. In one project I implemented monorepos with [lerna](https://github.com/lerna/lerna) and yarn workspaces. Scattering logic across multiple repos (or files in this case), increases the burden on the developer. This is especially true when the developer _needs_ to see related files to know how to figure out what to do.

There aren't unanimous best practices for separation of files. Mono vs many is still up for debate. PurelyFunctional disagrees with separating `subs` and `events` into separate files ([source](https://purelyfunctional.tv/guide/database-structure-in-re-frame/#events-and-subscriptions)). Some Clojurians have argued separating namespaces into `clj` `cljs` and `cljc` is bad practice. Here, I don't believe your points are well-formed enough to convince me otherwise.

Reply:
Thanks for your contribution on this issue - I love how you explained your coding style from your experience and was able to relate to other people from that experience. Now I see your point in full light, I wholeheartedly agree.

On another note, my mindset gave too much weight on other incoming "occasional & inexperienced" contributors over existing contributors. Ideally we'd like something with a literate programming paradigm so that end-user programming could be more powerful and integrated, but for now we can stick with what we have.

Regarding my upcoming contributions, I'll look into clojure metaprogramming first and try to implement some user-programmable components (after discussing the specs with you guys), then we could have more conversations regarding the backend. Kanbans & queries aren't a priority on my list because they're relatively independent and could make a great training ground for newly-graduated learners :)

Reply:
Kanbans are not a priority because I don't think anyone would miss them from Roam right now. Queries are in fact quite powerful, but might require more discussion from both eng. and design.

To your point on literate programming, mani and I briefly thought about this idea: [backup](https://backhub.co/) our github to an athens db and then begin using # and [[links]] in code comments. Link between issues, code, and docs seamlessly :)